cpp/ql/src/semmle/code/cpp/Function.qll

```diff
@@ -55,6 +56,48 @@ class Function extends Declaration, ControlFlowNode, AccessHolder, @function {
      )
   }
 
+  override string getIdentityString() {
+    result = getType().getTypeSpecifier() + getType().getDeclaratorPrefix() + " " + getScopePrefix(this) + getName() + getTemplateArgumentsString() + getDeclaratorSuffixBeforeQualifiers() + getDeclaratorSuffix()
+  }
+
+  language[monotonicAggregates]
+  private string getTemplateArgumentsString() {
```

Comment:
Since this duplicates part of `Function.getFullSignature`, can we have`getFullSignature` use this predicate?

Reply:
There would be subtly different output for `getFullSignature` for nested names if I made that change.

What I'd prefer to do is one of the following (in descending order of preference):
1. Remove `Function.getFullSignature()`. I believe I added it in 1.19, and it's only used in IR and AST dumps.
2. Rename `Symbol.getIdentityString()` to `System.getFullSignature()`, so that `Function.getFullSignature()` is just an override. `Function.getFullSignature()` would get the new output format, which should be OK since it's just meant for dumps.
3. Deprecate `Function.getFullSignature()` and leave it as-is for now.

Anyone have a preference?

Reply:
I don't see a reason to remove `getFullSignature` without deprecating it first.

If `getFullSignature` and `getIdentityString` aren't performant enough to be used on large snapshots, then I don't think they belong on `Function` and `Declaration` at all. No matter what we write in the QLDoc, users are going to call them anyway and have a bad experience. Is magic able to reduce what's computed in practice? For example, if someone writes `from Function f where matchesOnlyOneFunction(f) select f.getIdentityString()`, will the identity strings get computed for the whole snapshot or only for one function and its types?